File: verify/iscostas.go
package main

// TODO: documentation

import (
	"bufio"
	"fmt"
	"os"
)

func iscostas(perm []int) bool {
	n := len(perm)
	uniq := make(map[int]bool, n)
	for _, v := range perm {
		uniq[v] = true
	}
	if len(uniq) != n {
		return false
	}
	for k := 1; k < n; k++ {
		var seen uint64
		for i := 0; i < n-k; i++ {
			d := perm[i+k] - perm[i] + n
			bit := uint64(1) << d
			if seen&bit != 0 {
				return false
			}
			seen |= bit
		}
	}
	return true
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	var n int
	fmt.Fscan(reader, &n)
	perm := make([]int, n)
	for i := range perm {
		fmt.Fscan(reader, &perm[i])
	}
	if iscostas(perm) {
		fmt.Println(1)
	} else {
		fmt.Println(0)
	}
}

File: verify/iscostas.rs
use std::collections::HashSet;
use std::io::{self, BufRead};

// TODO: documentation

fn iscostas(perm: &[i64]) -> bool {
    let n = perm.len();
    let uniq: HashSet<_> = perm.iter().collect();
    if uniq.len() != n {
        return false;
    }
    for k in 1..n {
        let mut seen: u64 = 0;
        for i in 0..n - k {
            let d = (perm[i + k] - perm[i] + n as i64) as u32;
            let bit = 1u64 << d;
            if seen & bit != 0 {
                return false;
            }
            seen |= bit;
        }
    }
    true
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();
    let _n: usize = lines.next().unwrap().unwrap().trim().parse().unwrap();
    let perm: Vec<i64> = lines
        .next()
        .unwrap()
        .unwrap()
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect();
    println!("{}", if iscostas(&perm) { 1 } else { 0 });
}
